graspologic/plot/plot.py

```diff
@@ -671,6 +674,248 @@ def pairplot(
+        raise NameError(msg)
+    Y_, means, covariances = gmm.predict(X), gmm.means_, gmm.covariances_
+    data = pd.DataFrame(data=X)
+    n_components = gmm.n_components
+    # setting up the data DataFrame
+    if labels is None:
+        lab_names = [i for i in range(n_components)]
+        data["labels"] = np.asarray([lab_names[Y_[i]] for i in range(Y_.shape[0])])
+    else:
+        data["labels"] = labels
+    data["clusters"] = Y_
+    # creating a cluster_palette from a stirng  only if labels is given
+    if isinstance(cluster_palette, str) and labels is not None:
+        colors = sns.color_palette(cluster_palette, n_components)
+        cluster_palette = dict(zip(np.unique(Y_), colors))
+    # creating a label_palette from a stirng  only if labels is given
+    # note that these can be same string and mapping would be same for both
+    if isinstance(label_palette, str) and labels is not None:
+        colors = sns.color_palette(label_palette, n_components)
+        label_palette = dict(zip(np.unique(np.asarray(labels)), colors))
+    # must give labels AND palettes
+    if (labels is not None) and (cluster_palette is None or label_palette is None):
+        msg = "if give labels array must give palette dicts or string desciptors"
+        raise ValueError(msg)
+    elif labels is None and (cluster_palette is not None or label_palette is not None):
+        msg = "must give labels if using palettes for labels and clusters"
+        raise ValueError(msg)
+    # This takes care of default case, when no palettes or labels given
+    elif labels is None and cluster_palette is None and label_palette is None:
+        colors = sns.color_palette("Set1", n_components)
+        cluster_palette = dict(zip(np.unique(lab_names), colors))
+        label_palette = dict(zip(np.unique(lab_names), colors))
+
+    with sns.plotting_context(context=context, font_scale=font_scale):
+        # handle the case where only given two dims
+        if X.shape[1] == 2:
+            pairplot, axes = plt.subplots(1, 1, figsize=(12, 12))
+            _plot_ellipse(
+                data,
+                X,
+                0,
+                1,
+                means,
+                covariances,
+                axes,
+                label_palette,
+                cluster_palette,
+                alpha=alpha,
+            )
+            # formatting
+            axes.set_ylabel("Dim " + str(0))
+            axes.set_xlabel("Dim " + str(1))
+            axes.label_outer()
+            axes.spines["right"].set_visible(False)
+            axes.spines["top"].set_visible(False)
+            handles, labels = [], []
+            handles_, labels_ = axes.get_legend_handles_labels()
+            handles += handles_
+            labels += labels_
+            pairplot.legend(
+                handles[: means.shape[0] + 1],
+                labels[: means.shape[0] + 1],
+                loc="upper right",
+            )
```

Comment:
i'd still like to understand why all of this is necessary. it may be that there is a simpler solution, and if not, there should be a comment explaining why it is done like this.

Reply:
@bdpedigo  I do not think there is a simpler way, although i do admit i can make this a little more concise. The goal of this is to remove the repetitive information within the legend. I  will make this clear in the comments.

Reply:
https://github.com/microsoft/graspologic/pull/541/commits/d185ebb71a4db6c6e413b9e00e2beab9c6605f60

Reply:
@bdpedigo is this better?

Reply:
I think i address this here https://github.com/microsoft/graspologic/pull/541/commits/72d227c50777c95a9785ffe59d1cb292bae3b9b5